src/Libplanet.Types/Consensus/ValidatorSet.cs

```diff
@@ -264,16 +263,84 @@ public Validator GetProposer(long height, int round)
-        /// ordered, <see langword="false"/> otherwise.</returns>
-        public bool ValidateBlockCommitValidators(BlockCommit blockCommit)
+        /// <exception cref="InvalidBlockCommitException">Thrown when some votes in the
+        /// <paramref name="blockCommit"/> does not have non-null power.</exception>
+        /// <exception cref="InvalidBlockCommitException">Thrown when validators from
+        /// <paramref name="blockCommit"/> is different from validators of this.</exception>
+        public void ValidateBlockCommitValidators(BlockCommit blockCommit)
         {
-            return Validators.Select(validator => validator.PublicKey)
-                       .SequenceEqual(
-                           blockCommit.Votes.Select(vote => vote.ValidatorPublicKey).ToList()) &&
-                   blockCommit.Votes.All(
-                       v => v.ValidatorPower is null ||
-                            v.ValidatorPower == GetValidator(v.ValidatorPublicKey).Power);
+            ValidatorSet validatorSetFromCommit = new ValidatorSet(blockCommit.Votes.Select(
+                v => v.ValidatorPower is BigInteger power
+                    ? new Validator(v.ValidatorPublicKey, power)
+                    : throw new InvalidBlockCommitException(
+                        "All votes in the block commit after block protocol version 10 " +
+                        "must have power.")).ToList());
+
+            if (!Validators.Select(validator => validator.PublicKey)
+                    .SequenceEqual(
+                        blockCommit.Votes.Select(vote => vote.ValidatorPublicKey).ToList()))
+            {
+                throw new InvalidBlockCommitException(
+                    $"BlockCommit of BlockHash {blockCommit.BlockHash} " +
+                    $"has different validator set with chain state's validator set: \n" +
+                    $"in states | \n " +
+                    Validators.Aggregate(
+                        string.Empty, (s, v) => s + v.PublicKey + ", \n") +
+                    $"in blockCommit | \n " +
+                    blockCommit.Votes.Aggregate(
+                        string.Empty, (s, v) => s + v.ValidatorPublicKey + ", \n"));
+            }
+
+            if (!blockCommit.Votes.All(
+                v => v.ValidatorPower == GetValidator(v.ValidatorPublicKey).Power))
+            {
+                throw new InvalidBlockCommitException(
+                    $"BlockCommit of BlockHash {blockCommit.BlockHash} " +
+                    $"has different validator power with chain state's validator set: \n" +
+                    $"in states | \n " +
+                    Validators.Aggregate(
+                        string.Empty, (s, v) => s + v.Power + ", \n") +
+                    $"in blockCommit | \n " +
+                    blockCommit.Votes.Aggregate(
+                        string.Empty, (s, v) => s + v.ValidatorPower + ", \n"));
+            }
+        }
```

Comment:
Seems like merging this condition with above null condition will reduce 1 times of iteration.

Reply:
> Seems like merging this condition with above null condition will reduce 1 times of iteration.

It can be resolved by deleting the above code, rather than merging it with this code.

Reply:
Amended https://github.com/planetarium/libplanet/pull/3912/commits/5f5108e68967db43bea1070ac3a9e11aca5aacf4